kale/evaluate/uncertainty_metrics.py

```diff
@@ -0,0 +1,1036 @@
+                    if error <= upper and error > lower:
+                        inner_list_e.append(error)
+                        inner_list_id.append(id_)
+
+                else:
+                    lower = quantile_thresholds[q - 1]
+                    upper = 999999999999999999999999999999
+
+                    if error > lower:
+                        inner_list_e.append(error)
+                        inner_list_id.append(id_)
+
+            errors_groups.append(inner_list_e)
+            key_groups.append(inner_list_id)
+
+        # flip them so they go from B5 to B1
+        pred_bins_keys = pred_bins_keys[::-1]
+        pred_bins_errors = pred_bins_errors[::-1]
+        errors_groups = errors_groups[::-1]
+        key_groups = key_groups[::-1]
+
+        # Now for each bin, get the jaccard similarity
+        inner_jaccard_sims = []
+        inner_recalls = []
+        inner_precisions = []
+        for bin in range(num_bins):
+            pred_b_keys = pred_bins_keys[bin]
+            gt_bins_keys = key_groups[bin]
+
+            j_sim = jaccard_similarity(pred_b_keys, gt_bins_keys)
+            all_qs_jacc[bin].append(j_sim)
+            all_qs_jacc_concat_lms_sep[i][bin].append(j_sim)
+
+            inner_jaccard_sims.append(j_sim)
+
+            # If quantile threshold is the same as the last quantile threshold,
+            # the GT set is empty (rare, but can happen if distribution of errors is quite uniform).
+            if len(gt_bins_keys) == 0:
+                recall = 1.0
+                precision = 0.0
+            else:
+                recall = sum(el in gt_bins_keys for el in pred_b_keys) / len(gt_bins_keys)
+
+                if len(pred_b_keys) == 0 and len(gt_bins_keys) > 0:
+                    precision = 0.0
+                else:
+                    precision = sum(1 for x in pred_b_keys if x in gt_bins_keys) / len(pred_b_keys)
+
+            inner_recalls.append(recall)
+            inner_precisions.append(precision)
+            all_qs_recall[bin].append(recall)
+            all_qs_precision[bin].append(precision)
+
+        # print("Inner jaccard sims: ", inner_jaccard_sims)
+        all_lm_jacc.append(np.mean(inner_jaccard_sims))
+        all_lm_recall.append(np.mean(inner_recalls))
+        all_lm_precision.append(np.mean(inner_precisions))
+
+    return {
+        "mean all lms": np.mean(all_lm_jacc),
+        "mean all bins": [np.mean(x) for x in all_qs_jacc],
+        "all bins": all_qs_jacc,
+        "mean all lms recall": np.mean(all_lm_recall),
+        "mean all bins recall": [np.mean(x) for x in all_qs_recall],
+        "all bins recall": all_qs_recall,
+        "mean all lms precision": np.mean(all_lm_precision),
+        "mean all bins precision": [np.mean(x) for x in all_qs_precision],
+        "all bins precision": all_qs_precision,
+        "all bins concatenated lms seperated": all_qs_jacc_concat_lms_sep,
+    }
+
+
+def jaccard_similarity(list1: list, list2: list) -> float:
+    """
```

Comment:
Sure not available elsewhere?

Reply:
in discussion: https://github.com/pykale/pykale/discussions/392